File: day_06/part_two.py
def calculate_distance(hold_time: int, race_time: int) -> int:
    speed = hold_time
    remaining_time = race_time - hold_time

    return speed * remaining_time


def parse_numbers(line: str) -> int:
    split_str = line.split()
    number_str = ''.join(split_str[i] for i in range(1, len(split_str)))

    return int(number_str)


def part_two(input_file_path: str) -> int:
    with open(input_file_path) as file:
        lines = file.readlines()

        race_time = parse_numbers(lines[0])
        record_distance = parse_numbers(lines[1])

    possible_wins = 0

    for hold_time in range(0, race_time + 1):
        distance = calculate_distance(hold_time, race_time)

        if distance > record_distance:
            possible_wins += 1

    return possible_wins


if __name__ == "__main__":
    part_two_answer = part_two("input.txt")
    print("Part one:", part_two_answer)

File: Rust/day_06/src/part_two.rs
pub fn part_two(input_file_path: &str) -> usize {
    let lines = std::fs::read_to_string(input_file_path).expect("Failed to read file");
    let lines = lines.lines();

    let race_time = parse_numbers(lines.clone().nth(0).unwrap());
    let record_distance = parse_numbers(lines.last().unwrap());

    (0..=race_time).into_iter()
        .map(|hold_time| {calculate_distance(&hold_time, &race_time)})
        .filter(|distance| {*distance > record_distance})
        .count()
}

fn parse_numbers(line: &str) -> usize {
    let split_string = line.split_whitespace();

    split_string.skip(1)
        .map(|string| string.to_string())
        .collect::<Vec<String>>()
        .join::<&str>("")
        .parse::<usize>()
        .unwrap()
}

fn calculate_distance(hold_time: &usize, race_time: &usize) -> usize {
    let distance = hold_time * (race_time - hold_time);
    distance
} 

#[cfg(test)]
mod tests {
    use super::part_two;


    #[test]
    fn test_part_two() {
        let expected = 71503;
        let actual = part_two("test_input.txt");

        assert_eq!(expected, actual);
    }
}
